### ISSUE TITLE
Sum types?

### ISSUE THREAD

**Issue by Storyyeller** [stedolan/mlsub#6]
Are sum types supported? If so, would it be possible to document the syntax used in the demo?

**Comment by stedolan** [comment]
The online demo doesn't have sum type support. There's some experimental support for sum types in the master branch, but the syntax isn't exactly stable. At the moment, it looks like this:

```
def map(f, list) = match list {
  case 'nil => 'nil
  case 'cons(hd = x, tl = xs) => 'cons(hd = f(x), tl = map(f, xs))
}
```

**Comment by LPTK** [comment]
Are they implemented as polymorphic variants? The quotation mark before the name seems reminiscent of them.

**Comment by stedolan** [comment]
Yes, they're similar to OCaml's polymorphic variants in that the variant names are just labels, not defined identifiers. Internally they use subtyping, rather than the row variables used by OCaml.
